packages/cli/src/commands/build/index.ts

```diff
@@ -826,6 +832,51 @@ function mergeFlags(
   });
 }
 
+/**
+ * Takes the build output and writes all the variants into the `variants.json`
+ * file. It'll skip variants that already exist.
+ */
+async function writeVariantsJson(
+  { output }: Client,
+  buildResults: Iterable<BuildResult | BuildOutputConfig>,
+  outputDir: string
+): Promise<void> {
+  const variantsFilePath = join(outputDir, 'variants.json');
+
+  let hasVariants = true;
+
+  const variants = (await fs.readJSON(variantsFilePath).catch(error => {
+    if (error.code === 'ENOENT') {
+      hasVariants = false;
+      return { definitions: {} };
+    }
+
+    throw error;
+  })) as { definitions: Record<string, VariantDefinition> };
+
+  for (const result of buildResults) {
+    if (!('variants' in result) || !result.variants) continue;
+
+    for (const [key, defintion] of Object.entries(result.variants)) {
+      if (result.variants[key]) {
+        output.warn(
+          `The variant "${key}" was found multiple times. Only its first occurrence will be considered.`
+        );
+        continue;
+      }
+
+      hasVariants = true;
+      variants.definitions[key] = defintion;
+    }
+  }
+
+  // Only create the file when there are variants to write,
+  // or when the file already exists.
+  if (hasVariants) {
```

Comment:
<!-- Suggestion comments propose improvements to the code. Mention specifically what should be changed and what the benefit would be.  -->
**suggestion (non-blocking):** ‏remove need for `hasVariants`

```suggestion
  if (Object.keys(variants.definitions).length) {
```

Can we check if there are any variants in the data structure instead? As this code changes, it will be easy to forget to set `hasVariants` somewhere, but as long as `variants.definitions` is updated, this check will work.

Reply:
It would be possible now, but we might add more properties next to `definitions` that we'd need to check then. That's why the initial check is `true`.

Though, I'll see if I can improve it in follow up PRs.